File: apps/visualizer/backend/api/system.py
import logging
import os
import sys

import config
from flask import Blueprint, current_app, has_app_context, jsonify
from lightroom_tagger.core.database import init_database

logger = logging.getLogger(__name__)

# Add project root for lightroom_tagger imports
_PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
sys.path.insert(0, _PROJECT_ROOT)

bp = Blueprint('system', __name__)

@bp.route('/status', methods=['GET'])
def get_status():
    return jsonify({'status': 'ok'})

@bp.route('/vision-models', methods=['GET'])
def get_vision_models():
    """Get available vision models from all providers (registry + optional user DB rows)."""
    try:
        from lightroom_tagger.core.provider_registry import ProviderRegistry

        registry = ProviderRegistry()
        defaults = registry.defaults
        default_comparison = defaults.get("vision_comparison", {})
        default_provider = default_comparison.get("provider", "ollama")
        default_model = default_comparison.get("model")

        all_models = []
        for provider in registry.list_providers():
            if not provider["available"]:
                continue
            try:
                models = registry.list_models(provider["id"])
            except Exception:
                logger.exception("Failed to list models for provider %s", provider["id"])
                continue
            for model in models:
                if not model.get("vision"):
                    continue
                model_name = model["id"]
                is_default = (
                    provider["id"] == default_provider
                    and (default_model is None or default_model == model_name)
                    and not any(entry.get("default") for entry in all_models)
                )
                all_models.append({
                    "name": model_name,
                    "provider_id": provider["id"],
                    "default": is_default,
                })

        if has_app_context():
            db = getattr(current_app, "db", None)
            if db is not None:
                from database import get_user_models

                seen = {(entry["provider_id"], entry["name"]) for entry in all_models}
                for user_model in get_user_models(db):
                    if not user_model.get("vision"):
                        continue
                    key = (user_model["provider_id"], user_model["model_id"])
                    if key in seen:
                        continue
                    all_models.append({
                        "name": user_model["model_id"],
                        "provider_id": user_model["provider_id"],
                        "default": False,
                    })
                    seen.add(key)

        if not all_models:
            return jsonify({
                "models": [{"name": "gemma3:27b", "default": True}],
                "fallback": True,
            })

        if not any(entry["default"] for entry in all_models):
            all_models[0]["default"] = True

        return jsonify({"models": all_models, "fallback": False})
    except Exception:
        logger.exception("Failed to load vision models from registry")
        return jsonify({
            "models": [{"name": "gemma3:27b", "default": True}],
            "fallback": True,
        })

@bp.route('/stats', methods=['GET'])
def get_stats():
    try:
        db_path = config.LIBRARY_DB
        if not os.path.exists(db_path):
            return jsonify({'error': 'Library database not found'}), 404

        db = init_database(db_path)

        images = db.execute("SELECT * FROM images").fetchall()
        instagram_images = db.execute("SELECT * FROM instagram_dump_media").fetchall()

        posted_row = db.execute(
            "SELECT COUNT(*) AS cnt FROM images WHERE instagram_posted = 1"
        ).fetchone()
        posted_count = int(posted_row["cnt"])
        matches = db.execute("SELECT * FROM matches").fetchall()

        stats = {
            'catalog_images': len(images),
            'instagram_images': len(instagram_images),
            'posted_to_instagram': posted_count,
            'matches_found': len(matches),
            'db_path': db_path,
        }

        db.close()
        return jsonify(stats)
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@bp.route('/catalog/status', methods=['GET'])
def get_catalog_status():
    """Return whether the catalog vision cache has any prepared entries."""
    try:
        from lightroom_tagger.core.database import get_cache_stats

        db_path = config.LIBRARY_DB
        if not os.path.exists(db_path):
            return jsonify({'cached': False})

        db = init_database(db_path)
        try:
            stats = get_cache_stats(db)
        finally:
            db.close()
        return jsonify({'cached': stats['cached'] > 0})
    except Exception as e:
        return jsonify({'error': str(e)}), 500


_CACHE_PIPELINE_BUCKETS: tuple[tuple[str, str, dict], ...] = (
    # (bucket_key, job_type, extra_metadata_filter)
    # ``extra_metadata_filter`` keys map to ``json_extract(metadata, '$.<key>')``.
    # ``None`` means "absent or NULL" (legacy jobs without that key).
    ('catalog_sync', 'catalog_sync', {}),
    ('embed_catalog', 'batch_embed_image', {'image_type': ('catalog', None)}),
    ('embed_catalog_and_instagram', 'batch_embed_image', {'image_type': ('catalog_and_instagram',)}),
    ('stack_detect', 'batch_stack_detect', {}),
    ('catalog_similarity', 'batch_catalog_similarity', {}),
    ('catalog_cache_build', 'catalog_cache_build', {}),
    ('prepare_catalog', 'prepare_catalog', {}),
)


def _last_job_for_bucket(
    db,
    job_type: str,
    metadata_filter: dict[str, tuple],
) -> dict | None:
    """Return the most recent job row matching *job_type* + JSON metadata filter, or ``None``.

    Uses ``json_extract(metadata, '$.<key>')`` so we can split a single job_type
    (e.g. ``batch_embed_image``) into multiple UI buckets via metadata values.
    A ``None`` inside an allowed-values tuple matches rows where the key is
    absent (NULL) — preserves the historical "image_type defaults to catalog"
    semantics for legacy jobs.
    """
    clauses = ['type = ?']
    params: list = [job_type]
    for key, allowed in metadata_filter.items():
        col = f"json_extract(metadata, '$.{key}')"
        sub_clauses = []
        for value in allowed:
            if value is None:
                sub_clauses.append(f"{col} IS NULL")
            else:
                sub_clauses.append(f"{col} = ?")
                params.append(value)
        if sub_clauses:
            clauses.append('(' + ' OR '.join(sub_clauses) + ')')
    sql = (
        "SELECT id, type, status, created_at, started_at, completed_at, error "
        "FROM jobs WHERE " + ' AND '.join(clauses) +
        " ORDER BY created_at DESC LIMIT 1"
    )
    row = db.execute(sql, tuple(params)).fetchone()
    if row is None:
        return None
    row = dict(row)
    return {
        'job_id': row['id'],
        'type': row['type'],
        'status': row['status'],
        'created_at': row['created_at'],
        'started_at': row['started_at'],
        'completed_at': row['completed_at'],
        'error': row['error'],
    }


@bp.route('/cache/pipeline-status', methods=['GET'])
def get_cache_pipeline_status():
    """Latest run per Catalog Cache pipeline trigger.

    One entry per UI button on ``CatalogCacheTab``:

    * ``catalog_sync`` — most recent incremental catalog → library.db sync.
    * ``embed_catalog`` — most recent ``batch_embed_image`` with
      ``metadata.image_type`` of ``catalog`` or absent (legacy default).
    * ``embed_catalog_and_instagram`` — same but ``image_type`` is
      ``catalog_and_instagram``.
    * ``stack_detect`` — most recent ``batch_stack_detect``.
    * ``catalog_similarity`` — most recent ``batch_catalog_similarity``.
    * ``catalog_cache_build`` — most recent composite chain (D-08).
    * ``prepare_catalog`` — most recent pre-compress run.

    Each value is ``null`` when no matching job exists, otherwise:

    ``{ job_id, type, status, created_at, started_at, completed_at, error }``.
    """
    try:
        result: dict = {}
        for bucket_key, job_type, metadata_filter in _CACHE_PIPELINE_BUCKETS:
            result[bucket_key] = _last_job_for_bucket(
                current_app.db, job_type, metadata_filter,
            )
        return jsonify(result)
    except Exception as e:
        logger.exception('Failed to gather cache pipeline status')
        return jsonify({'error': str(e)}), 500


@bp.route('/cache/status', methods=['GET'])
def get_cache_status():
    """Get vision cache status."""
    try:
        from lightroom_tagger.core.config import load_config as load_lt_config
        from lightroom_tagger.core.database import get_cache_stats

        db_path = config.LIBRARY_DB
        if not os.path.exists(db_path):
            return jsonify({'error': 'Library database not found'}), 404

        db = init_database(db_path)

        cache_stats = get_cache_stats(db)

        # Use lightroom_tagger config for cache_dir
        lt_config = load_lt_config()
        cache_dir = lt_config.vision_cache_dir

        db.close()
        return jsonify({
            'total_images': cache_stats['total'],
            'cached_images': cache_stats['cached'],
            'missing': cache_stats['missing'],
            'cache_size_mb': round(cache_stats['cache_size_mb'], 2),
            'cache_dir': cache_dir,
        })
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: apps/visualizer/backend/src/api/system.ts
/**
 * System routes — health, stats, catalog cache readiness.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { existsSync } from 'node:fs';
import { config, loadLibraryConfig } from '../config.js';
import { ERROR_DB_NOT_FOUND } from '../constants/errors.js';
import { openDb, openLibraryDb } from '../db/connection.js';
import { getCachePipelineStatus } from '../db/jobs/pipeline-status.js';
import { getUserModels } from '../db/jobs/user-models.js';
import { getInsightsSummary } from '../db/library/insights.js';
import {
  getCacheStats,
  getImageCount,
  getPostedImagesCount,
  hasCachedEntries,
} from '../db/library/statistics.js';
import { ProviderRegistry } from '../providers/registry.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  CachePipelineStatus,
  CacheStatus,
  CatalogCacheReadyResponse,
  InsightsSummary,
  Stats,
  SystemStatusResponse,
  VisionModelsResponse,
} from './schemas/system.js';

export const systemRoutes = createOpenApiApp();

const statusRoute = createRoute({
  method: 'get',
  path: '/status',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(SystemStatusResponse) },
  }),
});

systemRoutes.openapi(statusRoute, (c) => c.json({ status: 'ok' }, 200));

const statsRoute = createRoute({
  method: 'get',
  path: '/stats',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Catalog statistics', content: jsonBody(Stats) },
    404: { description: 'Library database not found', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

systemRoutes.openapi(statsRoute, (c) => {
  const dbPath = config.LIBRARY_DB;
  if (!existsSync(dbPath)) {
    return c.json({ error: ERROR_DB_NOT_FOUND }, 404);
  }
  // Read-only: these routes never mutate, which also keeps them off the write
  // lock the job processor holds during a catalog sync.
  const db = openLibraryDb(dbPath, { readonly: true });
  try {
    return c.json(
      {
        // The Flask version did `len(get_all_images(db))`, materializing every row
        // to count them. COUNT(*) is the same number without loading the catalog.
        catalog_images: getImageCount(db),
        posted_to_instagram: getPostedImagesCount(db),
        db_path: dbPath,
      },
      200,
    );
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 500);
  } finally {
    db.close();
  }
});

const catalogStatusRoute = createRoute({
  method: 'get',
  path: '/catalog/status',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Catalog cache readiness', content: jsonBody(CatalogCacheReadyResponse) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

systemRoutes.openapi(catalogStatusRoute, (c) => {
  const dbPath = config.LIBRARY_DB;
  // A missing library DB is "not cached", not an error — matches api/system.py.
  if (!existsSync(dbPath)) return c.json({ cached: false }, 200);

  let db;
  try {
    db = openLibraryDb(dbPath, { readonly: true });
    return c.json({ cached: hasCachedEntries(db) }, 200);
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 500);
  } finally {
    db?.close();
  }
});

/**
 * These three do not use the `libraryDb` middleware.
 *
 * `/stats` and `/catalog/status` above already open their own connection, because
 * the system group is mounted at `/api` alongside every other group and a
 * `use('/*', …)` here would open a connection for requests it does not own.
 * `/cache/pipeline-status` reads a different database entirely.
 */

const insightsSummaryRoute = createRoute({
  method: 'get',
  path: '/insights-summary',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Insights tile counts', content: jsonBody(InsightsSummary) },
    404: { description: 'Library database not found', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

systemRoutes.openapi(insightsSummaryRoute, (c) => {
  const dbPath = config.LIBRARY_DB;
  if (!existsSync(dbPath)) return c.json({ error: ERROR_DB_NOT_FOUND }, 404);

  const db = openLibraryDb(dbPath, { readonly: true });
  try {
    return c.json(getInsightsSummary(db), 200);
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 500);
  } finally {
    db.close();
  }
});

const cacheStatusRoute = createRoute({
  method: 'get',
  path: '/cache/status',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Vision cache status', content: jsonBody(CacheStatus) },
    404: { description: 'Library database not found', content: jsonBody(ErrorBody) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

systemRoutes.openapi(cacheStatusRoute, (c) => {
  const dbPath = config.LIBRARY_DB;
  if (!existsSync(dbPath)) return c.json({ error: ERROR_DB_NOT_FOUND }, 404);

  const db = openLibraryDb(dbPath, { readonly: true });
  try {
    const cacheDir = loadLibraryConfig(config.LT_CONFIG_YAML).visionCacheDir;
    const stats = getCacheStats(db, cacheDir);
    return c.json(
      {
        ...stats,
        // Rounded to two places for display, as Flask did. The raw byte sum stays
        // in `getCacheStats` so a caller that wants precision can have it.
        cache_size_mb: Math.round(stats.cache_size_mb * 100) / 100,
      },
      200,
    );
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 500);
  } finally {
    db.close();
  }
});

const cachePipelineStatusRoute = createRoute({
  method: 'get',
  path: '/cache/pipeline-status',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Latest run per pipeline trigger', content: jsonBody(CachePipelineStatus) },
    500: { description: 'Server error', content: jsonBody(ErrorBody) },
  }),
});

systemRoutes.openapi(cachePipelineStatusRoute, (c) => {
  // The jobs table lives in `visualizer.db`. Flask reached it through
  // `current_app.db`, a connection opened once at startup and shared across
  // threads; opened per request here instead, which is what makes the route safe
  // to serve while the job processor is writing.
  const dbPath = config.VISUALIZER_DB;
  if (!existsSync(dbPath)) {
    // Flask would have raised on a missing app DB and answered 500; the message is
    // the one the caller can act on.
    return c.json({ error: `Jobs database not found: ${dbPath}` }, 500);
  }
  const db = openDb(dbPath, { readonly: true, vec: false });
  try {
    return c.json(getCachePipelineStatus(db), 200);
  } catch (e) {
    return c.json({ error: e instanceof Error ? e.message : String(e) }, 500);
  } finally {
    db.close();
  }
});

// --- vision models ----------------------------------------------------------

const visionModelsRoute = createRoute({
  method: 'get',
  path: '/vision-models',
  tags: ['system'],
  responses: withValidationError({
    200: { description: 'Vision-capable models', content: jsonBody(VisionModelsResponse) },
  }),
});

/**
 * Vision-capable models from every available provider, plus user-added rows.
 *
 * Never fails: any error — an unreadable `providers.json`, a provider that throws
 * while listing — falls back to `gemma3:27b` with `fallback: true`, so the
 * description UI always has something selectable. That is deliberate in the
 * original and kept here; a 500 would leave the page with an empty dropdown and no
 * explanation.
 */
systemRoutes.openapi(visionModelsRoute, async (c) => {
  // Not `as const`: the response type needs a mutable array of the schema's shape.
  const fallbackBody = (): z.infer<typeof VisionModelsResponse> => ({
    models: [{ name: 'gemma3:27b', default: true }],
    fallback: true,
  });
  try {
    const registry = new ProviderRegistry();
    const descriptionDefaults = registry.defaults.description;
    const defaultProvider = descriptionDefaults?.provider ?? 'ollama';
    const defaultModel = descriptionDefaults?.model ?? null;

    const allModels: { name: string; provider_id: string; default: boolean }[] = [];
    for (const provider of await registry.listProviders()) {
      if (!provider.available) continue;
      let models;
      try {
        models = await registry.listModels(provider.id);
      } catch {
        // One broken provider must not empty the whole dropdown.
        continue;
      }
      for (const model of models) {
        if (!model.vision) continue;
        // The first match for the configured default wins, and only one entry is
        // ever marked — hence the check against what has already been added.
        const isDefault =
          provider.id === defaultProvider &&
          (defaultModel === null || defaultModel === model.id) &&
          !allModels.some((entry) => entry.default);
        allModels.push({ name: model.id, provider_id: provider.id, default: isDefault });
      }
    }

    if (existsSync(config.VISUALIZER_DB)) {
      const db = openDb(config.VISUALIZER_DB, { readonly: true, vec: false });
      try {
        const seen = new Set(allModels.map((entry) => `${entry.provider_id}\u0000${entry.name}`));
        for (const userModel of getUserModels(db)) {
          if (!userModel.vision) continue;
          const key = `${userModel.provider_id}\u0000${userModel.model_id}`;
          if (seen.has(key)) continue;
          seen.add(key);
          allModels.push({
            name: userModel.model_id,
            provider_id: userModel.provider_id,
            default: false,
          });
        }
      } finally {
        db.close();
      }
    }

    if (allModels.length === 0) return c.json(fallbackBody(), 200);
    // Something must be selected, even when no provider matched the default.
    if (!allModels.some((entry) => entry.default)) allModels[0]!.default = true;
    return c.json({ models: allModels, fallback: false }, 200);
  } catch {
    return c.json(fallbackBody(), 200);
  }
});
